src/fheroes2/heroes/heroes_action.cpp

```diff
@@ -2591,7 +2591,7 @@ void ActionToUpgradeArmyObject( Heroes & hero, const MP2::MapObjectType objectTy
 
             // border
             fheroes2::Blit( border, surface, offsetX, offsetY );
-            fheroes2::renderMonsterFrame( *mons[i], surface, { 6 + offsetX, 6 + offsetY } );
+            fheroes2::renderMonsterFrame( ( *mons[i] ).GetUpgrade(), surface, { 6 + offsetX, 6 + offsetY } );
```

Comment:
While this syntax for dereferencing a pointer is valid, there is a less clunky one:

```suggestion
            fheroes2::renderMonsterFrame( mons[i]->GetUpgrade(), surface, { 6 + offsetX, 6 + offsetY } );
```